### ISSUE TITLE
Its not convient (or possible) to use checkpointing in pytorch because masks is not tensor type objects

### ISSUE THREAD

**Issue by hadaev8** [idiap/fast-transformers#21]

**Comment by angeloskath** [comment]
Hi, could you elaborate a bit? The masks are not part of the module, they are input to the module.

Would it be possible to provide an example to recreate your issue?

**Comment by hadaev8** [comment]
Sorry for the delay.
Basically, I can use it easily without masks like this:

```
        builder = TransformerEncoderBuilder.from_kwargs(
            n_layers=depth,
            n_heads=n_heads,
            query_dimensions=base//n_heads,
            value_dimensions=base//n_heads,
            feed_forward_dimensions=base*4
        )
        builder.attention_type = "linear"
        self.attn = builder.get()
```

and in forward
`x = torch.utils.checkpoint.checkpoint(self.attn, x)`
But it accepts only tensors as arguments.
So I need to make a wrapper, pass x tensor and lengths tensor, and make masks inside.

**Comment by angeloskath** [comment]
I am in the process of fixing this and I was wondering what would you prefer, the ability to pass tensors as masks or a simple checkpoint wrapper function?

I like the wrapper function idea because it will be able to best preserve the input of masks as objects. Having objects for masking is important because different attention mechanisms require different style of masking.

In fact, there is a probability to add an extra arguments object that might contain additional arguments for different transformers (which we can also account for when using a wrapper function). For instance Longformer requires the global attention tokens to be passed together with the input.

I am leaning towards the following:

```python
from fast_transformers.utils import checkpoint

transformer = TransformerEncoderBuilder().get()
....
x = checkpoint(transformer, x, attn_mask=attn_mask)
```

Let me know if you feel strongly against such a solution.

Cheers,
Angelos

**Comment by hadaev8** [comment]
How flexible it will be?
I want to make 2 encoder 1 decoder model in near future.
Would be silly to make my own alteration for this module.
